sherpa/astro/data.py

```diff
@@ -839,7 +839,7 @@ def ignore_bad(self):
 
     def _dynamic_group(self, group_func, *args, **kwargs):
 
-        keys = kwargs.keys()[:]
+        keys = list(kwargs.keys())[:]
```

Comment:
I believe `list` returns a copy of the iterable, so we can drop the trailing `[:]` from such cases. See https://docs.python.org/3/library/stdtypes.html#list
(this is in a number of places throughout the code base).